academic-services/knowledge-exploration-service/reference-grammar-syntax.md

```diff
@@ -0,0 +1,225 @@
+
+```xml
+  <one-of>
+    <item>before</item>
+    <item logprob="-1">written before</item>
+  </one-of>
+```
+
+### ruleref element
+
+The `ruleref` element specifies valid expansions via references to another `rule` element.  Through the use of `ruleref` elements, more complex expressions can be built from simpler rules.  The required `uri` attribute indicates the name of the referenced `rule` using the syntax "#*rulename*".  To capture the semantic output of the referenced rule, use the optional `name` attribute to specify the name of a variable to which the semantic output is assigned.
+ 
+```xml
+<ruleref uri="#paperYear" name="matchedAttribute" />
+```
+
+### attrref element
+
+The `attrref` element references an index attribute, allowing matching against attribute values observed in the index.  The required `uri` attribute specifies the index schema name and attribute name using the syntax "*schemaName*#*attrName*".  There must be a preceding `import` element that imports the schema named *schemaName*.  The attribute name is the name of an attribute defined in the corresponding schema.
+
+In addition to matching user input, the `attrref` element also returns a structured query object as output that selects the subset of objects in the index matching the input value.  Use the optional `name` attribute to specify the name of the variable where the query object output should be stored.  The query object can be composed with other query objects to form more complex expressions.  See [structured query expressions](concepts-query-expressions.md) for details.  
+
+```xml
+<attrref uri="paperEntity#C.CN" name="matchedAttribute" />
+```
+
+### tag element
+
+The `tag` element specifies how a path through the grammar is to be interpreted.  It contains a sequence of semicolon-terminated statements.  A statement may be an assignment of a literal or a variable to another variable.  It may also assign the output of a function with 0 or more parameters to a variable.  Each function parameter may be specified using a literal or a variable.  If the function does not return any output, the assignment is omitted.  Variable scope is local to the containing rule.
+
+```xml
+<tag>x = 1; y = x;</tag>
+<tag>q = All(); q = And(q, q2);</tag>
+<tag>AssertEquals(x, 1);</tag>
+```
+
+Each `rule` in the grammar has a predefined variable named "out", representing the semantic output of the rule.  Its value is computed by evaluating each of the semantic statements traversed by the path through the `rule` matching the user query input.  The value assigned to the "out" variable at the end of the evaluation is the semantic output of the rule.  The semantic output of interpreting a user query against the grammar is the semantic output of the root rule.
```

Comment:
>Each [](start = 0, length = 4)

How about adding "### out variable" before line 89?

Reply:
Great catch!

Reply:
Restructured that general area a bit